File: hermes/src/main/java/com/lodz/android/hermes/modules/WsClient.java
package com.lodz.android.hermes.modules;

import android.os.Looper;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.drafts.Draft;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URI;
import java.util.Map;

/**
 * WebSocket客户端
 * @author zhouL
 * @date 2019/5/7
 */
public class WsClient extends WebSocketClient {

    private OnWebSocketListener mListener;

    public WsClient(String url) {
        super(URI.create(url));
    }

    public WsClient(URI serverUri, Draft protocolDraft) {
        super(serverUri, protocolDraft);
    }

    public WsClient(URI serverUri, Map<String, String> httpHeaders) {
        super(serverUri, httpHeaders);
    }

    public WsClient(URI serverUri, Draft protocolDraft, Map<String, String> httpHeaders) {
        super(serverUri, protocolDraft, httpHeaders);
    }

    public WsClient(URI serverUri, Draft protocolDraft, Map<String, String> httpHeaders, int connectTimeout) {
        super(serverUri, protocolDraft, httpHeaders, connectTimeout);
    }

    @Override
    public void onOpen(ServerHandshake handshakedata) {
        doOpen(handshakedata);
    }

    @Override
    public void onMessage(String message) {
        doMessage(message);
    }

    @Override
    public void onClose(int code, String reason, boolean remote) {
        doClose(code, reason, remote);
    }

    @Override
    public void onError(Exception ex) {
        doError(ex);
    }

    @Override
    public void connect() {
        try {
            super.connect();
        }catch (Exception e){
            e.printStackTrace();
            doError(e);
        }
    }

    @Override
    public void reconnect() {
        Thread thread = new Thread(this);
        thread.start();
        super.reconnect();
    }

    /** 当前是否在主线程（UI线程） */
    private boolean isMainThread() {
        return Looper.myLooper() == Looper.getMainLooper();
    }

    private void doOpen(final ServerHandshake handshakedata) {
        UiHandler.post(new Runnable() {
            @Override
            public void run() {
                if (mListener != null) {
                    mListener.onOpen(handshakedata);
                }
            }
        });
    }

    private void doMessage(final String message) {
        UiHandler.post(new Runnable() {
            @Override
            public void run() {
                if (mListener != null) {
                    mListener.onMessage(message);
                }
            }
        });
    }

    private void doClose(final int code, final String reason, final boolean remote) {
        UiHandler.post(new Runnable() {
            @Override
            public void run() {
                if (mListener != null) {
                    mListener.onClose(code, reason, remote);
                }
            }
        });
    }

    private void doError(final Exception ex) {
        UiHandler.post(new Runnable() {
            @Override
            public void run() {
                if (mListener != null) {
                    mListener.onError(ex);
                }
            }
        });
    }

    /**
     * 设置监听器
     * @param listener 监听器
     */
    public void setOnWebSocketListener(OnWebSocketListener listener){
        mListener = listener;
    }

    public interface OnWebSocketListener {

        void onOpen(ServerHandshake handshakedata);

        void onMessage(String message);

        void onClose(int code, String reason, boolean remote);

        void onError(Exception e);
    }
}

File: hermes/src/main/java/com/lodz/android/hermes/modules/WsClient.kt
package com.lodz.android.hermes.modules

import com.lodz.android.hermes.contract.OnWebSocketListener
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.java_websocket.client.WebSocketClient
import org.java_websocket.handshake.ServerHandshake
import java.net.SocketException
import java.net.URI

/**
 * WebSocket客户端
 * @author zhouL
 * @date 2019/12/21
 */
class WsClient(url: String) : WebSocketClient(URI.create(url)) {

    private var mListener: OnWebSocketListener? = null

    override fun onOpen(handshakedata: ServerHandshake?) {
        doOpen(handshakedata)
    }

    override fun onClose(code: Int, reason: String?, remote: Boolean) {
        doClose(code, reason ?: "", remote)
    }

    override fun onMessage(message: String?) {
        doMessage(message ?: "")
    }

    override fun onError(ex: Exception?) {
        doError(ex ?: SocketException("exception is null"))
    }

    override fun connect() {
        try {
            super.connect()
        } catch (e: Exception) {
            e.printStackTrace()
            doError(e)
        }
    }

    override fun reconnect() {
        val thread = Thread(this)
        thread.start()
        super.reconnect()
    }

    private fun doOpen(handshakedata: ServerHandshake?) {
        GlobalScope.launch(Dispatchers.Main) {
            mListener?.onOpen(handshakedata)
        }
    }

    private fun doMessage(message: String) {
        GlobalScope.launch(Dispatchers.Main) {
            mListener?.onMessage(message)
        }
    }

    private fun doClose(code: Int, reason: String, remote: Boolean) {
        GlobalScope.launch(Dispatchers.Main) {
            mListener?.onClose(code, reason, remote)
        }
    }


    private fun doError(e: Exception) {
        GlobalScope.launch(Dispatchers.Main) {
            mListener?.onError(e)
        }
    }

    /** 设置监听器[listener] */
    fun setOnWebSocketListener(listener: OnWebSocketListener) {
        mListener = listener
    }
}